### PR TITLE
Block the creation of empty commits

### PR DESCRIPTION
**Pull request by pjbgf**
In specific cases, the update mechanism may lead to no files being changed. The changes short-circuit the process to exit early when 0 files have been modified.

go-git has been changed to return an error when an empty commit is detected, which is used as a last resort.

When the push branch is different than the clone branch, the controller will now download all branch references from the Git server, in order to enable a single fetch operation.

Fixes https://github.com/fluxcd/image-automation-controller/issues/466.

### CODE REVIEW

File: controllers/imageupdateautomation_controller.go [hunk 1]
```diff
@@ -358,21 +366,28 @@ func (r *ImageUpdateAutomationReconciler) Reconcile(ctx context.Context, req ctr
...
-			Message: message,
-		},
-		repository.WithSigner(signingEntity),
-	); err != nil {
-		if err != git.ErrNoStagedFiles {
+			repository.WithSigner(signingEntity),
+		)
+	} else {
+		err = extgogit.ErrEmptyCommit
+	}
```
Comment (aryan9600) [inline controllers/imageupdateautomation_controller.go:387]: why are we creating an error here?

Comment (pjbgf) [reply controllers/imageupdateautomation_controller.go:387]: This is to be able to reuse the existing logic for handling empty commits.

Given that this controller will be refactored as part of the GA process, I found this was the easiest way to resolve the problem without duplicating code.

**Review by aryan9600 — APPROVED** [review]
lgtm! 🚀

**Review by hiddeco — APPROVED** [review]
Thank you @pjbgf :100:

**Review by stefanprodan — APPROVED** [review]

**Pull request merged**